src/composition_root.rs

```diff
@@ -1,34 +1,46 @@
 use crate::networking::{TcnApiImpl, TcnApi};
 use crate::reports_updater::{TcnMatcher, ReportsUpdater, TcnDao, TcnDaoImpl, TcnMatcherImpl};
-use crate::{reporting::{memo::{MemoMapperImpl}, symptom_inputs::{SymptomInputsSubmitterImpl, SymptomInputsSubmitter}}, preferences::{Preferences, PreferencesImpl}, tcn_ext::tcn_keys::{TcnKeysImpl, TcnKeys}};
+use crate::{reporting::{memo::MemoMapperImpl, symptom_inputs::{SymptomInputsSubmitterImpl, SymptomInputsSubmitter, SymptomInputs}, symptom_inputs_manager::{SymptomInputsManagerImpl, SymptomInputsProcessorImpl, SymptomInputsProcessor}}, preferences::{Preferences, PreferencesImpl}, tcn_ext::tcn_keys::TcnKeysImpl};
 use once_cell::sync::Lazy;
 use parking_lot::RwLock;
 use std::sync::Arc;
 
-pub struct CompositionRoot<'a,
-  PreferencesType: Preferences, TcnDaoType: TcnDao, TcnMatcherType: TcnMatcher, ApiType: TcnApi, 
-  // TODO don't pass concrete type for MemoMapper / TcnKeys here?
-  SymptomInputsSubmitterType: SymptomInputsSubmitter<MemoMapperImpl, TcnKeysImpl<PreferencesType>, ApiType>, 
-> {
-  pub api: ApiType,
-  pub reports_updater: ReportsUpdater<'a, PreferencesType, TcnDaoType, TcnMatcherType, ApiType>,
-  pub symptom_inputs_submitter: SymptomInputsSubmitterType
+pub struct CompositionRoot<'a, A, B, C, D, E, F> where 
```

Comment:
This is a very complicated data type. Does it make sense to split it up?

Reply:
Well, it's the composition root (aka dependency graph). It's not different than what we use in iOS and Android. Rust has the "problem" that it needs to know the actual types(sizes) at compile time, so the generics are more verbose. Otherwise we have to use `Box` or similar, I think. I agree that the generics are unwieldy, though. Maybe there's a better solution.